src/functions/artifacts.ts

```diff
@@ -182,3 +204,100 @@ export function hasSameCleanupPolicy(
 export function hasCleanupOptOut(repo: artifactregistry.Repository): boolean {
   return !!(repo.labels && repo.labels[OPT_OUT_LABEL_KEY] === "true");
 }
+
+/**
+ * Checks whether a clean up policy is required for Artifact Registry in given locations.
+ */
+export async function checkCleanupPolicy(
+  projectId: string,
+  locations: string[],
+): Promise<{ locationsToSetup: string[]; locationsWithErrors: string[] }> {
+  if (locations.length === 0) {
+    return { locationsToSetup: [], locationsWithErrors: [] };
+  }
+
+  const checkRepos = await Promise.allSettled(
+    locations.map(async (location) => {
+      try {
+        const repository = await exports.getRepo(projectId, location);
+        const hasPolicy = !!findExistingPolicy(repository);
+        const hasOptOut = hasCleanupOptOut(repository);
+        const hasOtherPolicies =
+          repository.cleanupPolicies &&
+          Object.keys(repository.cleanupPolicies).some((key) => key !== CLEANUP_POLICY_ID);
+
+        return {
+          location,
+          repository,
+          hasPolicy,
+          hasOptOut,
+          hasOtherPolicies,
+        };
+      } catch (err) {
+        logger.debug(`Failed to check artifact cleanup policy for region ${location}:`, err);
+        throw err;
+      }
+    }),
+  );
+
+  const locationsToSetup = [];
+  const locationsWithErrors = [];
+
+  for (let i = 0; i < checkRepos.length; i++) {
+    const result = checkRepos[i];
+    if (result.status === "fulfilled") {
+      if (!(result.value.hasPolicy || result.value.hasOptOut || result.value.hasOtherPolicies)) {
+        locationsToSetup.push(result.value.location);
+      }
+    } else {
+      locationsWithErrors.push(locations[i]);
+    }
+  }
+  return { locationsToSetup, locationsWithErrors };
+}
+
+/**
+ * Sets Artifact Registry cleanup policies for given locations.
+ */
+export async function setCleanupPolicies(
+  projectId: string,
+  locations: string[],
+  daysToKeep: number,
+): Promise<{ locationsWithPolicy: string[]; locationsWithErrors: string[] }> {
+  if (locations.length === 0) return { locationsWithPolicy: [], locationsWithErrors: [] };
+
+  const locationsWithPolicy: string[] = [];
+  const locationsWithErrors: string[] = [];
+
+  const setupRepos = await Promise.allSettled(
+    locations.map(async (location) => {
+      try {
+        logger.debug(`Setting up artifact cleanup policy for region ${location}`);
+        const repo = await exports.getRepo(projectId, location);
+        await exports.setCleanupPolicy(repo, daysToKeep);
+        return { region: location, success: true };
+      } catch (err) {
+        logger.debug(`Failed to set up artifact cleanup policy for region ${location}:`, err);
+        return { region: location, success: false, error: err };
+      }
+    }),
+  );
+
+  for (const result of setupRepos) {
+    if (result.status === "rejected") {
+      // should not happen
```

Comment:
Nit: I'm personally a fan of throwing if it should never happen (can use status 1 to mean "Firebase screwed up" in telemetry)

Reply:
Alternatively, just use the promise settled result instead of 'success" and "err"